### ISSUE TITLE
Alternate audio tracks

### ISSUE THREAD

**Issue by drequivalent** [Chocobozzz/PeerTube#939]
What if Peertube videos supported alternate audio tracks? Like, the way we have subtitles?

Or even better, audio track blending!

And also, the suggestion mechanism for audio tracks.

It would be very useful for dub translations into another languages. This feature is badly missing from YouTube, which results in a lot of unnecessary duplication of the same video with very varying video quality.

If we have this feature, we always get the exact same video, but we can choose an audio track we prefer, or even have it selected automatically according to browser locale. Great for watching documentaries and open movies.

**Comment by ealgase** [comment]
Please add this, it would be amazing!

**Comment by buoyantair** [comment]
Interesting concept indeed :O

**Comment by StalinEXE** [comment]
+1
I really want to create some videos for English and Russian speaking auditory (I really ready to dub video twice in these two languages). It is a really cool feature that will help to avoid creating two different videos in two languages.

Just open the video and click on English or Russian in interface. I like this idea and hope to see it in PeerTube.

**Comment by bkil** [comment]
It would be interesting to consider what you would set the `language` metadata field to? Maybe convert it to a list? That could confuse video search engines, though.

**Comment by zsolt-beringer** [comment]
@bkil A list or set would be a good solution. Every search query containing a `language` field related filtering that matches any of the listed audio languages of a video in the set would match.

**Comment by bkil** [comment]
@zsolt-beringer This is how the field currently looks like:

```
  "data": [
    {
      "privacy": {
        "id": 1,
        "label": "Public"
      },
      "language": {
        "id": "hu",
        "label": "Hungarian"
      },
      ...
```

I think if we didn't want to break downstream users, it would be wise to keep this as is to convey the default language and add a new key (like `languages` or `audioTracks`) that contains a list. I wouldn't necessarily limit ourselves to using a json map. I could imagine wanting to create two different audio tracks in the same language, for example one containing the original noisy recording made at a conference and a second one rerecorded in a studio environment that is less hurtful to the ears.

**Comment by bkil** [comment]
Surely a proper solution would be a bit costly, although I think it is worth it. This could come either by storing the tracks separately and combining them in the player (#1448), or implementing a much more complicated bitstream parser/demultiplexer as well. I haven't looked at the code, but I have a feeling this would take a while.

I could imagine a bit hacky, but cheap minimal viable solution until then. For example, we could just transcode and store the given entire video file multiple times (once per audio track), and the player would choose one file based on the required audio track. In the vast majority of use cases, people usually upload content in at most two languages, hence the overhead isn't that excessive.

Even if an optimized backend would be implemented at a later time, most of the user facing features would still work afterwards. The video's link would be stable, description and comments would be combined, content recommendation would stay unified. Though, the comment section made me wondering - is it worthwhile at all to cram comment together from different languages below a given video?

Or would the end goal be more motivated more by wanting to share video streams to save storage costs (and to improve seeding)? If the end goal is more about seeding, it would improve the swarm a lot if all languages were kept in the same file (and/or all of them would be downloaded together).

As a less redundant, but much more hacky workaround, the video file could be stored once with the default audio track, and the alternate audio tracks could be stored in separate files. Now comes the kludge: if a user selects a different audio track, the player would mute the main file and start playing the alternate audio track in the background, keeping the playback of the muted video and the separate audio in sync. Sounds brittle, but it might just work.

**Comment by mmuman** [comment]
For the record, most common A/V container formats (MP4, AVI…) handle multiple audio tracks with language IDs just fine. But supporting only this way means all the tracks we want must already be in the initial file.

**Comment by kontrollanten** [comment]
Another positive upside with separating audio from video would be that playing audio in background on mobile would work. Since HLS supports alternate audio tracks there's no need to care about syncing the tracks, and there's still possible to download the videos since most video players supporting playing m3u8 files.

**Comment by rigelk** [comment]
> In Video.js, _we only allow one track to be enabled at a time_; so, if you enable more than one, the last one to be enabled will end up being the only one. While the spec allows for more than one track to be enabled, Safari and most implementations only allow one audio track to be enabled at a time.

source: https://docs.videojs.com/tutorial-audio-tracks.html

That means audio track blending won't be a thing for us any time soon, unless the change happens upstream at videojs.

**Comment by kontrollanten** [comment]
> > In Video.js, _we only allow one track to be enabled at a time_; so, if you enable more than one, the last one to be enabled will end up being the only one. While the spec allows for more than one track to be enabled, Safari and most implementations only allow one audio track to be enabled at a time.
> 
> source: https://docs.videojs.com/tutorial-audio-tracks.html
> 
> That means audio track blending won't be a thing for us any time soon, unless the change happens upstream at videojs.

I understand that as they're talking about _audio tracks_, not tracks in general:

> Audio tracks are a feature of HTML5 video for providing alternate audio track selections to the user, so that a track other than the main track can be played.

According to this demo it seems to work with audio and video tracks enabled at the same time https://codepen.io/team/rcrooks1969/pen/LoWNQy

**Comment by rigelk** [comment]
> > That means audio track blending won't be a thing for us any time soon, unless the change happens upstream at videojs.
> 
> I understand that as they're talking about _audio tracks_, not tracks in general:

I'm talking about audio tracks too ; more specifically _audio track blending_ (meaning trying to have more than one audio track enabled at once).

**Comment by bkil** [comment]
I don't think that this issue involves audio track blending. What we would need is for the user to be able to choose an audio track in their own language, preferably chosen automatically by browser language or PeerTube account preferences after first load.

**Comment by rigelk** [comment]
@bkil I'm merely answering the second line of the issue description.

**Comment by NoMoreCRAPTion** [comment]
> I hope this gets added! YouTube does not have this feature in any way. It would help so many create dubs and even audio description tracks, rather than uploading multiple videos.

YouTube now has this feature. :)

https://www.tubefilter.com/2021/04/12/youtube-testing-language-dubbing-tool-audio-tracks/

"How YouTube's new audio feature is helping Ubisoft make its trailers more accessible" 

https://www.techradar.com/news/how-youtubes-new-audio-feature-is-helping-ubisoft-make-its-trailers-more-accessible

"YouTube Testing Language Dubbing Tool, Enabling Viewers To Toggle Between Multiple ‘Audio Tracks’"

https://www.tubefilter.com/2021/04/12/youtube-testing-language-dubbing-tool-audio-tracks/

**Comment by rkingett** [comment]
This is great YouTube added it. Even so, this could allow for creators to do many different things with their channel.

**Comment by whyronimus** [comment]
I only can support the idea given here. It's no problem to include multiple audio tracks in video files. 

Two more suggestions:

- It should be possible to specify the labels of alternate audio tracks, e.g. "Stereo", "Headhone Stereo", "5.1" when uploading or updating a video
- as indicated, other formats than stereo should be possible

**Comment by Slender1808** [comment]
> I only can support the idea given here. It's no problem to include multiple audio tracks in video files.
> 
> Two more suggestions:
> 
> * It should be possible to specify the labels of alternate audio tracks, e.g. "Stereo", "Headhone Stereo", "5.1" when uploading or updating a video
> * as indicated, other formats than stereo should be possible

I believe the labels depend on the video file type.

**Comment by whyronimus** [comment]
> I believe the labels depend on the video file type.

The publishing form could read out how many audio streams are in the file, and save one label per audio stream which should be able to be chosen in the player. I am thinking it as the label being saved in the database, not the video file (I don't even know if this is possible)

**Comment by whyronimus** [comment]
@Chocobozzz could you give any information if this feature has been discussed or added to the roadmap?

It should not be difficult to implement, as far as I understood the videojs player is used. Tracks could be read from the video file with `ffprobe`, and added to the player as described [here](https://docs.videojs.com/tutorial-audio-tracks.html) or [here](https://github.com/videojs/videojs-contrib-hls/blob/master/docs/multiple-alternative-audio-tracks.md). Ideally, `ffprobe` should read out the metadata such as `title` from the audio track and propose it in the publishing form, but it should be editable.

I haven't found any solution for Audio Track blending or simultaneous playback, but this should not be necessary IMHO.

**Comment by Slender1808** [comment]
would it be possible to use tar file with video separated more inside the same tar?

**Comment by njohng** [comment]
Would like to see this being implemented, with the ability to add multiple audio tracks to an existing video at any time. YouTube allows background tracks to be added to a published video. A similar implementation would be ideal.

 Videos [here](https://www.nitansolutions.com/multilingual_videos.html) have multiple audio tracks as translations. A Video.js based player? .. not sure.

**Comment by whyronimus** [comment]
@njohng good idea, this would be possible with the `-map` command for ffmpeg.
If anyone here is willing to give me some basic guidance, I could try to write a plugin for it. Never done before, but willing.

**Comment by Agorise** [comment]
This is BADLY needed. I should be able to switch to different subtitle languages as easily as switching the audio to a different language as well. I swear I will dump my dvd collection once Peertube has this feature :)

**Comment by stevespaw** [comment]
This is related to this issue - and ROKU compatibility. If it were possible to adjust the FFMPEG output options like the Encoding plugin can do for input/transcode properties, we may be able to get there. If we can get this  Multi Audio hurdle done, Peertube could really dominate some projects! 
[https://github.com/Chocobozzz/PeerTube/issues/2625#issuecomment-1548080453](url)

**Comment by uis246** [comment]
> Or would the end goal be more motivated more by wanting to share video streams to save storage costs (and to improve seeding)? If the end goal is more about seeding, it would improve the swarm a lot if all languages were kept in the same file (and/or all of them would be downloaded together).

I hope decoder can selectively fetch needed audio stream only.

> Now comes the kludge: if a user selects a different audio track, the player would mute the main file and start playing the alternate audio track in the background, keeping the playback of the muted video and the separate audio in sync. Sounds brittle, but it might just work.

Sounds like average webdev hack. Just get current timestamp decoded and start playing another language from that timestamp. Assuming audio file has samplerate and integer offset from zero in samples it will be simple stopping decoding initial audio stream and starting decoding new audio stream from offset(keep in mind seeking sometimes can be expensive). If audio streams have integer samplerate and seconds-aligned offset, then wait until next second starts and replace tracks.

**Comment by dannekrose** [comment]
I’d like to add my support and bump this issue. Is there anything the average person can do to help?

**Comment by stevespaw** [comment]
This is the Roku problem that is also related - Fragmented MP4, CMAF (muxing audio and video not supported for CMAF)..  https://developer.roku.com/docs/specs/media/streaming-specifications.md

**Comment by FediVideos** [comment]
Does the latest 6.3.0 update which separates video and audio allow multiple audio tracks on a single video? Asking as there's a group which does audio description for video who would like to try doing this on PeerTube, and multiple audio would allow this to happen.

(For those who don't know, audio description is where there's an alternate audio track where the on-screen action is described for people with visual impairments but the dialogue is left intact.)

**Comment by uis246** [comment]
> (For those who don't know, audio description is where there's an alternate audio track where the on-screen action is described for people with visual impairments but the dialogue is left intact.)

This is even more important than localization!

**Comment by rkingett** [comment]
[Here is a sample of audio description](https://www.youtube.com/watch?v=iXmCFHDvlzA)

**Comment by candidexmedia** [comment]
Here's an example of how the AblePlayer video player handles video descriptions: https://ableplayer.github.io/ableplayer/demos/desc1.html

**Comment by Chocobozzz** [comment]
> Does the latest 6.3.0 update which separates video and audio allow multiple audio tracks on a single video?

Not yet, but it was a necessary step if we implement alternate audio tracks in the future

**Comment by nerzhul** [comment]
Hello, thanks for this amazing project first.
Please note that support would be nice, especially when we import videos from youtube. Now they have multiple tracks, it select english by default it seems, instead of native one

**Comment by JohnySeven** [comment]
Hello,
this feature would be great to implement!

Also, I think Elena Rossini would second that: https://mastodon.social/@_elena/114715096535543365 :-)

Thank you for great product!
Jan

**Comment by eleit** [comment]
> I hope this gets added! YouTube does not have this feature in any way. It would help so many create dubs and even audio description tracks, rather than uploading multiple videos.

+1 for this feature. I self-host my own PeerTube instance on a small VPS and it would be so much more effective / economical (from the POV of storage) to have the ability to add foreign language tracks to a video, instead of doing standalone foreign versions. 

Plus having foreign videos that people aren't interested in federating and popping up in people's feeds is not ideal...

**Comment by bkil** [comment]
Ironically, this issue aged badly and is becoming less and less relevant. By today, it is possible to render the mouth of each speaker using DeepFake-like image processing methods so they appear to be originally speaking in the given alternate language corresponding to the track. You would want to upload separate video tracks for each language in this case.

**Comment by mmuman** [comment]
Ethical and ecological issues of AI notwithstanding, the cost of hosting videos is already still prohibitive to some, having to host 10 times the same video just for a few pixels of difference seems just ridiculous when just adding audio tracks would be good enough.

**Comment by ulfi** [comment]
> > I hope this gets added! YouTube does not have this feature in any way. It would help so many create dubs and even audio description tracks, rather than uploading multiple videos.
> 
> +1 for this feature. I self-host my own PeerTube instance on a small VPS and it would be so much more effective / economical (from the POV of storage) to have the ability to add foreign language tracks to a video, instead of doing standalone foreign versions.
> 
> Plus having foreign videos that people aren't interested in federating and popping up in people's feeds is not ideal...

+1 for this feature.

**Comment by schmaker** [comment]
> Ironically, this issue aged badly and is becoming less and less relevant. By today, it is possible to render the mouth of each speaker using DeepFake-like image processing methods so they appear to be originally speaking in the given alternate language corresponding to the track. You would want to upload separate video tracks for each language in this case.

You actually need both, but audio should have to be higher priority because of more use cases.

I can imagine to have multiple video tracks when (for example) video has localized text included in itself, so you could create one video with multiple languages video track

Multiple audio tracks are needed for localization, commentary tracks, accessibility tracks, ..

It's also lighter for servers to have multiple audio tracks instead of having one video per language

**Comment by candidexmedia** [comment]
> Ethical and ecological issues of AI notwithstanding, the cost of hosting videos is already still prohibitive to some, having to host 10 times the same video just for a few pixels of difference seems just ridiculous when just adding audio tracks would be good enough.

> You actually need both, but audio should have to be higher priority because of more use cases.

Agreed, plus not everyone wants some weird looking AI-generated mouth movements and robotic voices. In some industries, that kind of thing would bring the integrity and truth of the footage into question.

Multiple audio-streams are already suppported by some video container formats, and they offer some pretty interesting possibillities (ex: a version of the video with background music vs no music).

**Comment by bkil** [comment]
Accessibility tracks sounds like an odd use case to list. The visual impaired would prefer audio narration version of a movie served with 0px video, hence it would need to be a completely new upload anyway.

Many high budget movies already use such image processing techniques.

Funny how startups succeeded in brainwashing laypeople into calling anything at all "AI" nowadays. You could substitute "AI" everywhere in any text to "something" and it would still not lose information. There was a fascinating movie about this in 2006.

**Comment by bkil** [comment]
By the way, you usually do not hear "robotic voices" in such movies. They are still narrated by professional voice actors. The difference is that the lip and the throat is in sync with the text thanks to this, without having to film every scene 100x times with 100x actors. I don't see what other choice you have to solve this problem.

https://en.wikipedia.org/wiki/McGurk_effect

Discounting its utility feels like an ignorant viewpoint from someone who never watches foreign movies and who does not believe that anyone speaks a language other than English.

**Comment by uis246** [comment]
> Ironically, this issue aged badly and is becoming less and less relevant. By today, it is possible to render the mouth of each speaker using DeepFake-like image processing methods so they appear to be originally speaking in the given alternate language corresponding to the track. You would want to upload separate video tracks for each language in this case.

Because you can do it doesn't mean you should do it. Unless you propose ability to match any video track with any audio track like VLC does, your comment is meaningless.

Also stating this should not be implemented because throwing brute-forced statistics at video is easy and cheap nowdays is self-contradictory. If it is so easy, then why not let client choose model they like?

**Comment by schmaker** [comment]
> Accessibility tracks sounds like an odd use case to list. The visual impaired would prefer audio narration version of a movie served with 0px video, hence **it would need to be a completely new upload** anyway.

That's wrong .. as audio track is already a separated file, all you need for this is adding "sound only" quality to the video

**Comment by kamazeuci** [comment]
+1 for this feature... unless it has already been implemented?

**Comment by Simx72** [comment]
This is a very necessary feature in small instances, because making a copy of each video takes space, when you only need to add a new audio track, not the whole video. It could even be presented as separate videos, but internally use the same video file and different audio file.

**Comment by gavio-dot** [comment]
It would be a really useful feature

**Comment by mmuman** [comment]
I tried some months ago to add an audio track on youtube and found out it was reserved to some channels… So it'd also be a differentiating feature for a lot of people.

**Comment by drakonicguy** [comment]
+1. I make films and videos in multiple languages, it would be neat to have them be just one multilanguage upload.

**Comment by rkingett** [comment]
As a screen reader user, YouTube had this for a while now, but there is a major problem at least for me the way YouTube has done it. There is no way to tell how many audio tracks are in a video without playing the video. With PeerTube, I'd like an indicator or text letting me know how many audio tracks a video has before I begin to play the video or even far before I review the video details page, like if there is a list of videos, I'd like to know if it has CC and how many audio tracks it has before going to the videos details page.

**Comment by Simx72** [comment]
> I'd like to know if it has CC and how many audio tracks it has before going to the videos details page.

I vote for this. Maybe even have a user setting say 'preferred languages' that you want, and then the videos preview will have badges (or a list for screen readers) showing your preferred languages if the video includes them, and maybe a flag for the language.

**Comment by AlchemiistCreative** [comment]
+1. This would be very useful for multilingual companies that rebroadcast internal meetings. And since HLS already supports splitting audio and video streams, that sounds like something that could be possible.

**Comment by omartek** [comment]
+1 Asbsolute necessary 🙏
